### ISSUE TITLE
ENH: rec.fromrecords accept dict

### ISSUE THREAD

**Issue by toddrjen** [numpy/numpy#5934]
It is currently relatively complicated to create a structured or record array from a `dict`.  This simplest approach I can find, `numpy.rec.fromrecords`, still requires a fair amount of boilerplate:

    data={1: [1,2,3], 2:[4,5,6]}
    np.rec.fromrecords(list(data.values()), names=list(data.keys()))

I suggest allowing `numpy.rec.fromrecords` to allow using a dict directly:

    data={1: [1,2,3], 2:[4,5,6]}
    np.rec.fromrecords(data)

Alternatively, there could be separate convenience functions to construct record arrays and structured arrays from dicts or sequences of key/value pairs.

**Comment by roger-** [comment]
+1

Would be great if it accepted both dicts and lists of dicts, like pandas.

**Comment by ahaldane** [comment]
I'm interested in such improvements too.

Something that worries me here is that the key-order in dicts is not preserved by Python, which means that the layout of the structure/record arrays you get may be somewhat random. In `fromrecords({1: [1,2,3], 2:[4,5,6]})`, you might sometimes end up with field `1` at byte-offset 0, and other times it will be field `2`. I would prefer if you gave a series of tuples, which preserves order. If you started from a dict and didn't care about field order, you could still do

    np.rec.fromrecords(data.iteritems())

Also, for

    data2 = [{'c': 7, 'd': 8}, {'e': 9, 'f': 10}]

what is the intended result?

**Comment by roger-** [comment]
> If you started from a dict and didn't care about field order...

True, and you can also use an OrderedDict to get around the ordering issue. Or maybe there could be a `fieldorder` parameter to specify the order of the keys?

> what is the intended result?

Sorry, fixed my example so there's only `c`/`d`. Should be more obvious now.

Ultimately I think storing data in dicts / lists of dicts is common enough that it should be directly supported to make recarrays more accessible.

**Comment by toddrjen** [comment]
> Something that worries me here is that the key-order in dicts is not preserved by Python, which means that the layout of the structure/record arrays you get may be somewhat random.

Yes, if you want to preserve order, you will need another approach, such as sorting or OrderedDict.  This doesn't seem to be a problem for most projects, however.

> In `fromrecords({1: [1,2,3], 2:[4,5,6]})`, you might sometimes end up with field `1` at byte-offset 0, and other times it will be field `2`.

Technically, the order of elements of a dict is undefined, but still deterministic (at least in CPython, in Pypy all dicts are ordered).  So your example will always result in the same thing.  

> I would prefer if you gave a series of tuples, which preserves order. If you started from a dict and didn't care about field order, you could still do

I don't see why these are mutually-exclusive.  I agree that supporting a series of tuples is also a good idea, but considering how common dicts are in Python I think it is important to support those as well.

**Comment by roger-** [comment]
I propose something that works like this:

```python
from numpy.core.records import fromrecords

def fromdict(data, *args, **kwargs):
    if isinstance(data, dict):
        return fromrecords(zip(*data.values()), *args, names=data.keys(), **kwargs)
        
    if isinstance(data[0], dict):
        keys = data[0].keys()
        vals = [tuple(d[k] for k in keys) for d in data]
        
        return fromrecords(vals, *args, names=keys, **kwargs)

    raise Exception('input must be dict or list of dicts')
```

Test:

```python
data1 = {'x': [1,2,3], 'y': [10, 20, 30]}
data2 = [{'x': 1, 'y': 10}, {'x': 2, 'y': 20}, {'x': 3, 'y': 30}]

fromdict(data1)
>>> rec.array([(10, 1), (20, 2), (30, 3)], 
      dtype=[('y', '<i4'), ('x', '<i4')])

fromdict(data2)
>>> rec.array([(10, 1), (20, 2), (30, 3)], 
      dtype=[('y', '<i4'), ('x', '<i4')])
```

Pandas probably has a more efficient way of implementing this.

**Comment by anntzer** [comment]
I would also appreciate a dict-to-structured array constructor (basically, similar to the DataFrame constructor taking a dict).
The originally raised issue of nondeterministic dict iteration order no longer applies since Py3.6.  Assuming numpy follows NEP29 itself (does it?), Py3.5 support can be dropped by June 2020, which (based on the recent release rhythm) would correspond to numpy 1.19?

**Comment by charris** [comment]
>Py3.5 support can be dropped by June 2020

Python 3.5 has been dropped in master, so 1.19 it will be.
